Improve HeapBlockDevice, TDBStore and tests

LDong-Arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Fixes: #14128, #14157

The PR does the following:
* In #14007 we removed the use of the shim layer `FlashSimBlockDevice` from TDBStore tests and passed the real `HeapBlockDevice` directly to TDBStore, because flash simulation was no longer needed after #11987. This caused TDBStore tests to _fail_ (instead of _skip_ properly) on low-RAM targets, because `HeapBlockDevice::erase()` does _not_ trigger any heap allocations, unlike `FlashSimBlockDevice::erase()` which we used before.
To fix this issue, use `HeapBlockDevice::program()` instead which allocates heap memory on the fly for blocks being programmed. If this fails, skip the test for not having enough heap.
* A few quality-of-life improvements:
  * `HeapBlockDevice`: Use `new (std::nothrow)` for all allocations, instead of a mixture of `malloc` and unchecked `new` (exceptions are not enabled by Mbed OS).
  * `HeapBlockDevice::erase()`: Deallocate memory of erased blocks.
  * `FileSystemStore` unit test: add missing reformat step after erase.
  * Optimize `TDBStore`'s erase operations. It also always calls `BlockDevice::erase()` regardless of erase value, as erase-before-program is required by `BlockDevice`'s documentation.

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

None.

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
Manual testing on K64F:
* Full RAM used: TDBStore tests run and pass, as before
* RAM size reduced from 0x30000 to 0x6000 in the linker file (to emulate a low-RAM target): TDBStore tests are skipped correctly.
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

@ARMmbed/mbed-os-core @geky @davidsaada 

----------------------------------------------------------------------------------------------------------------

LDong-Arm committed:
HeapBlockDevice: allocate with std::nothrow

Mbed OS does not enable C++ exceptions, so we should call `new` with
`std::nothrow` which returns a C-style NULL pointer when allocation
fails to allow error handling.

For consistency of style within the same file, this commit also
replaces `malloc()` and `free()` to `new (std::nothrow)` and `delete`.

LDong-Arm committed:
HeapBlockDevice::erase(): free up heap memory

`HeapBlockDevice::erase()` previously performed a range and alignment
check only. This commit adds freeing of heap memory.

LDong-Arm committed:
TDBStore Whitebox tests: fix memory check

Each block of HeapBlockDevice is only allocated from the heap when
that block is programmed. And erasing a block frees the associated
buffer.

To decide if there is enough heap to run the TDBStore Whitebox tests,
we need to perform a trial program() instead of erase().

LDong-Arm commented on storage/blockdevice/include/blockdevice/HeapBlockDevice.h:
```diff
@@ -93,8 +95,6 @@ class HeapBlockDevice : public BlockDevice {
 
     /** Program blocks to a block device
-     *
-     *  The blocks must have been erased prior to being programmed
```
I removed this because it's definitely not true - the heap buffer is only allocated when a block is programmed, not when it's erased.

Patater replied:
I read this as being about checks the class method will do to enforce that the block is erased before being programmed; policy you must follow as a programmer. I don't see code to check this, but maybe there ought to be some code here if we want to ensure programs written to use the HeapBlockDevice will continue to work when used with actual flash-based devices.

I checked the implementation of `is_valid_program()`, but it doesn't appear to check that a block was erased before it can be programmed. If this policy is enforced anywhere, I'm not clear on where yet.

LDong-Arm replied:
Good point. I think it's a real inconsistency in our existing API.

Mbed OS allows a `BlockDevice` to be either flash-based or not.
* For flash-based ones (e.g. `SPIFBlockDevice`), I think the actual erase requirement is set by the hardware (if I understand correctly).
* For non-flash-based ones, "erase" should in theory be unsupported. One example is SD. From [its documentation](https://os.mbed.com/docs/mbed-os/v6.9/apis/sdblockdevice.html)
    > Because an SD card doesn't need preparation, erase is not applicable on SDBlockDevice; it's a no-op operation.

So there's no single policy for all `BlockDevice`s.

`HeapBlockDevice` has always been quite self-contradicting:
* It was _intended_ to be non-flash-based. This was why we used to use a flash simulation layer [`FlashSimBlockDevice`](https://github.com/ARMmbed/mbed-os/blob/master/storage/blockdevice/source/FlashSimBlockDevice.cpp) on top of it, to simulate erase operations when used for `TDBStore` which used to required erase (but not anymore).
* However, a partially implemented & partially lacking `HeapBlockDevice::erase()` always existed - it performs size/alignment check, but doesn't do the actual erase.
* And this line of Doxygen we are talking about suggests erase is required before program.

In my opinion, the proper solution for best consistency is remove `::erase` and any mentions of "erase" from `HeapBlockDevice`. This would change the API for users who do like to use `::erase`.

Any preference?

LDong-Arm replied:
In the SD example I used above, its Doxygen contradicts with its documentation.
Documentation:
> Because an SD card doesn't need preparation, erase is not applicable on SDBlockDevice; it's a no-op operation.

Doxygen:
https://github.com/ARMmbed/mbed-os/blob/8075b2348cf8250a99f47eb410ad42e1ef2eea6a/storage/blockdevice/COMPONENT_SD/include/SD/SDBlockDevice.h#L122-L122

So HeapBlockDevice is just another one.

Patater replied:
Could you update SDBlockDevice too, please?

If we remove erase from HeapBlockDevice, it might no longer be swappable for other BlockDevice implementations. HeapBlockDevice::erase() may need to be a no-op for compatibility reasons (with perhaps things like LittleFS layered on top). Also, if HeapBlockDevice::program() allocates memory, what should free the allocated memory? In an earlier iteration of this PR, you had HeapBlockDevice::erase() free memory. If HeapBlockDevice::erase() is instead a no-op, should we move the freeing of allocated memory to HeapBlockDevice::~HeapBlockDevice() instead?

LDong-Arm replied:
The base class's virtual `BlockDevice::erase()` is already a no-op. The freeing of allocated memory is also done in `HeapBlockDevice::~HeapBlockDevice()` already. So removing `HeapBlockDevice::erase()` and updating Doxygen should be enough.

LDong-Arm replied:
Following discussions below, I kept `HeapBlockDevice::erase()` because it doesn't break compatibility/consistency with the definition of the `BlockDevice` API.

LDong-Arm committed:
TDBStore: optimize erase algorithms

Currently `TDBStore::offset_in_erase_unit()` and
`TDBStore::check_erase_before_write()` loop through erase units
one-by-one, until the entire range is covered. This is very inefficient
when the erase size is tiny, e.g. one-byte on a non-flash device for
which we use program as erase.

This commit reworks the algorithms, based on the fact that a block
device can erase or program as many units as needed in one go.

LDong-Arm committed:
FileSystemStore unit test: remove erased_set_get test

The "erased_set_get" test case deinits `FileSystemStore`, erases the
block device, and reinits `FileSystemStore`. This of course fails,
because `BlockDevice::erase()` removes all stored data including the
format of the `FileSystem` (middle layer), unless the particular type
of block device has a no-op erase implementation.

Note: Previously `HeapBlockDevice::erase()` was essentially a no-op
so this test case did not fail. We recently added the freeing of heap
memory and it uncovered the problem.

LDong-Arm commented on storage/kvstore/filesystemstore/tests/UNITTESTS/FileSystemStore/moduletest.cpp:
```diff
@@ -73,6 +73,9 @@ TEST_F(FileSystemStoreModuleTest, erased_set_get)
     EXPECT_EQ(heap.deinit(), MBED_SUCCESS);
+    if (fs->mount(&heap) != MBED_SUCCESS) {
+        EXPECT_EQ(fs->reformat(&heap), MBED_SUCCESS);
+    }
```
See https://github.com/ARMmbed/mbed-os/pull/12652#discussion_r605544000

LDong-Arm replied:
TODO: Remove this test case as per https://github.com/ARMmbed/mbed-os/pull/12652#discussion_r612145136

LDong-Arm committed:
TDBStore: call `BlockDevice::erase()` regarless of erase value

From the documentations of `BlockDevice::get_erase_value()`:

    -1 if you can't rely on the value of the erased storage

and `BlockDevice::program()`:

    The blocks must have been erased prior to being programmed

So, `BlockDevice::erase()` should always be called regardless of
erase value.
